src/main/scala/com/scality/clueso/MetadataIngestionPipeline.scala

```diff
@@ -0,0 +1,187 @@
+      recordNo
+    } else {
+      recordNo + compactionRecordInterval - recordNo % compactionRecordInterval
+    }
+  }
+
+  import org.apache.spark.sql.functions.udf
+  val msg_rewrite = udf(msgRewriteFun)
+  val max_op_index = udf(maxOpIndexFun)
+
+  /**
+    * Applies projections and conditions to incoming data frame
+    * Ignores malformed json and garbage payloads
+    * Ignores events from specified bucket name
+    *
+    * @param bucketNameToFilterOut
+    * @param eventStream
+    * @return
+    */
+  def filterAndParseEvents(bucketNameToFilterOut : String, eventStream: DataFrame)(implicit spark : SparkSession, config: CluesoConfig) = {
+
+    var df = eventStream.select(
+      col("timestamp").cast(TimestampType).as("kafkaTimestamp"),
+      msg_rewrite(col("value")).as("content")
+    )
+
+    // defensive filtering to not process kafka garbage
+    df = df.filter(col("content").isNotNull)
+          .filter(length(col("content")).gt(3))
+          .select(
+            col("kafkaTimestamp"),
+            from_json(col("content"), CluesoConstants.eventSchema).alias("event")
+          )
+
+    df = df.filter(col("event").isNotNull.and(col("event.type").isNotNull))
+      .withColumn("key", when(
+        col("event").isNotNull.and(
+          col("event.key").isNotNull
+        ), col("event.key")).otherwise("")
+      )
+      .withColumn("bucket",
+        when(
+          col("event").isNotNull.and(
+            col("event.bucket").isNotNull
+          ), col("event.bucket")).otherwise("NOBUCKET")
+      )
+      .withColumn("type", col("event.type"))
+      .withColumn("opIndex", col("event.opIndex"))
+      .withColumn("maxOpIndex", max_op_index(lit(config.compactionRecordInterval), col("event.opIndex")))
+      .withColumn("message", col("event.value"))
+
+    df = df.filter(
+      !col("bucket").eqNullSafe(bucketNameToFilterOut) &&
+        !col("bucket").eqNullSafe("users..bucket") &&
+        !col("bucket").eqNullSafe("__metastore") &&
+        (col("bucket").isNotNull && !col("bucket").startsWith("mpuShadowBucket")))
+
+      df.drop("event")
+  }
+
+  def main(args: Array[String]): Unit = {
+    require(args.length > 0, "specify configuration file")
+
+    implicit val config = SparkUtils.loadCluesoConfig(args.head)
+
+    // create dir no matter what
+    val fs = SparkUtils.buildHadoopFs(config)
+
+    logger.info(s"Creating directory ${PathUtils.landingURI}")
+    logger.info(s"Creating directory ${PathUtils.stagingURI}")
+
+    fs.mkdirs(new Path(PathUtils.landingURI))
+    fs.mkdirs(new Path(PathUtils.stagingURI))
+
+
+    implicit val spark = SparkUtils.buildSparkSession(config)
+      .appName("Metadata Ingestion Pipeline")
+      .getOrCreate()
+
+    val eventStream = spark.readStream
+      .format("kafka")
+      .option("kafka.bootstrap.servers", config.kafkaBootstrapServers)
+      .option("subscribe", config.kafkaTopic)
+      .option("startingOffsets", "earliest")
```

Comment:
is this the safest mode? so we might have duplicates but we'll always get the metadata?

Reply:
Exactly – although checkpointing is ON, which helps recovery, if something goes wrong we are pessimistically saying that we're willing to process all the records if needs be and have duplicates – the searches would have to crunch more results, but we know that logic is solid and de-dups before returning results